msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java

```diff
@@ -0,0 +1,184 @@
+        interactiveRequestIntent.putExtra(
+                AuthenticationConstants.Broker.BROKER_REQUEST_V2,
+                msalBrokerRequestAdapter.brokerRequestFromAcquireTokenParameters(parameters)
+        );
+
+        //Pass this intent to the BrokerActivity which will be used to start this activity
+        final Intent brokerActivityIntent = new Intent(parameters.getAppContext(), BrokerActivity.class);
+        brokerActivityIntent.putExtra(BrokerActivity.BROKER_INTENT, interactiveRequestIntent);
+
+        //Start the BrokerActivity
+        parameters.getActivity().startActivity(brokerActivityIntent);
+
+        //Wait to be notified of the result being returned... we could add a timeout here if we want to
+        final Bundle resultBundle = mBrokerResultFuture.get();
+
+        return getAcquireTokenResult(resultBundle);
+
+    }
+
+    /**
+     * Get the intent for the broker interactive request
+     *
+     * @param request
+     * @return
+     */
+    private Intent getBrokerAuthorizationIntent(@NonNull final AcquireTokenOperationParameters request) throws ClientException {
+        Intent interactiveRequestIntent;
+        IMicrosoftAuthService service;
+
+        final MicrosoftAuthClient client = new MicrosoftAuthClient(request.getAppContext());
+        final MicrosoftAuthServiceFuture authServiceFuture = client.connect();
+
+        try {
+            service = authServiceFuture.get();
+            interactiveRequestIntent = service.getIntentForInteractiveRequest();
+
+            // TODO : See what's the right exception here
+        } catch (RemoteException e) {
+            throw new RuntimeException("Exception occurred while attempting to invoke remote service", e);
+        } catch (Exception e) {
+            throw new RuntimeException("Exception occurred while awaiting (get) return of MicrosoftAuthService", e);
+        } finally {
+            client.disconnect();
+        }
+
+        return interactiveRequestIntent;
+    }
+
+    /**
+     * Get the response from the Broker captured by BrokerActivity.
+     * BrokerActivity will pass along the response to the broker controller
+     * The Broker controller will map th response into the broker result
+     * And signal the future with the broker result to unblock the request.
+     *
+     * @param requestCode
+     * @param resultCode
+     * @param data
+     */
+    @Override
+    public void completeAcquireToken(int requestCode, int resultCode, Intent data) {
+        mBrokerResultFuture.setResultBundle(data.getExtras());
+    }
+
+    @Override
+    public AcquireTokenResult acquireTokenSilent(AcquireTokenSilentOperationParameters parameters) throws BaseException {
+        IMicrosoftAuthService service;
+
+        MicrosoftAuthClient client = new MicrosoftAuthClient(parameters.getAppContext());
+        MicrosoftAuthServiceFuture future = client.connect();
+
+        try {
+            //Do we want a time out here?
+            service = future.get();
```

Comment:
Yeap, and we could call MicrosoftAuthServiceFuture.get(long l, TimeUnit timeUnit) instead to set the timeout.

Reply:
Yes the question is what's the time limit though, I will create another work item to discuss and add this we needed